File: data-structures/linked-lists/Go/LinkedList.go
package main

import (
	"fmt"
)

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (list *LinkedList) append(data int) {
	if list.head == nil {
		list.head = &Node{data, nil}
		return
	}

	current := list.head
	for current.next != nil {
		current = current.next
	}
	current.next = &Node{data, nil}
}

func (list *LinkedList) prepend(data int) {
	list.head = &Node{data, list.head}
}

func (list *LinkedList) deleteWithValue(data int) {
	if list.head == nil {
		return
	}

	if list.head.data == data {
		list.head = list.head.next
	}

	current := list.head
	for current.next != nil {
		if current.next.data == data {
			current.next = current.next.next
			return
		}
		current = current.next
	}
}

func (list *LinkedList) display() {
	if list.head == nil {
		fmt.Println("[]")
		return
	}

	current := list.head
	fmt.Println("[")
	for current != nil {
		fmt.Printf("\t%d\n", current.data)
		current = current.next
	}
	fmt.Println("]")
}

func main() {
	// Create list
	var list LinkedList
	list.display()

	// Append
	fmt.Println("---------- Append ----------")
	list.append(42)
	list.display()

	list.append(51)
	list.display()

	list.append(11)
	list.append(455)
	list.display()

	// Prepend
	list = LinkedList{}
	fmt.Println("---------- Prepend ----------")
	list.prepend(42)
	list.display()

	list.prepend(51)
	list.display()

	list.prepend(11)
	list.prepend(455)
	list.display()

	// Delete with value
	list = LinkedList{}
	fmt.Println("---------- Delete with value ----------")
	list.append(42)
	list.append(51)
	list.append(11)
	list.append(455)

	list.deleteWithValue(11)
	list.display()

	list.deleteWithValue(455)
	list.display()

	list.deleteWithValue(42)
	list.display()
}

File: data-structures/linked-lists/Java/LinkedList.java
// "static void main" must be defined in a public class.
public class SinglyLinkedList {
    public class Node {
        int data;
        Node next = null;

        public Node(int data) {
            this.data = data;
        }
    }

    Node head;

    public void append(int data) {
        if (head == null) {
            head = new Node(data);
            return;
        }

        Node current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = new Node(data);
    }

    public void prepend(int data) {
        Node newHead = new Node(data);
        newHead.next = head;
        head = newHead;
    }

    public void deleteWithValue(int data) {
        if (head == null) {
            return;
        }

        if (head.data == data) {
            head = head.next;
        }

        Node current = head;
        while (current.next != null) {
            if (current.next.data == data) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    public void display() {
        Node current = head;

        if(head == null) {
            System.out.println("[]");
            return;
        }

        System.out.println("[");
        while(current != null) {
            System.out.println("\t" + current.data);
            current = current.next;
        }
        System.out.println("]");
    }

    public static void main(String[] args) {
        // Create list
        SinglyLinkedList list = new SinglyLinkedList();
        list.display();

        // Append
        System.out.println("---------- Append ----------");
        list.append(42);
        list.display();

        list.append(51);
        list.display();

        list.append(11);
        list.append(455);
        list.display();

        // Prepend
        list = new SinglyLinkedList();
        System.out.println("---------- Prepend ----------");
        list.prepend(42);
        list.display();

        list.prepend(51);
        list.display();

        list.prepend(11);
        list.prepend(455);
        list.display();

        // Delete with value
        list = new SinglyLinkedList();
        System.out.println("---------- Delete with value ----------");
        list.append(42);
        list.append(51);
        list.append(11);
        list.append(455);

        list.deleteWithValue(11);
        list.display();

        list.deleteWithValue(455);
        list.display();

        list.deleteWithValue(42);
        list.display();
    }
}
